### PR TITLE
Documentation in src

### PR DESCRIPTION
**Pull request by jdkandersson**
- Add documentation upload action: https://github.com/canonical/upload-charm-docs
- Copy documentation into `docs`

### CODE REVIEW

File: metadata.yaml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 name: indico
 display-name: Indico
 summary: Indico web application.
-docs: https://discourse.charmhub.io/t/indico-docs-index/6552
```
Comment (arturo-seijas) [inline metadata.yaml:4]: Why this change?

**Review by gregory-schiano — DISMISSED** [review]
LGTM

File: docs/how-to-guides/contributing.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  ](https://github.com/canonical/indico-operator/issues) explaining your use case.
+- If you would like to chat with us about your use-cases or proposed implementation, you can reach
+  us at [Canonical Mattermost public channel](https://chat.charmhub.io/charmhub/channels/charm-dev)
+  or [Discourse](https://discourse.charmhub.io/).
+- Familiarising yourself with the [Charmed Operator Framework](https://juju.is/docs/sdk) library
+  will help you a lot when working on new features or bug fixes.
+- All enhancements require review before being merged. Code review typically examines
+  - code quality
+  - test coverage
+  - user experience for Juju administrators of this charm.
```
Comment (arturo-seijas) [inline docs/how-to-guides/contributing.md:17]: I'd go with `operators`rather than `administrators`

File: docs/how-to-guides/contributing.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+```
+
+### Testing
+
+```shell
+tox -e fmt           # update your code according to linting rules
+tox -e lint          # code style
+tox -e unit          # unit tests
+# integration tests (see below for building docker images)
+tox -e integration -- --indico-image localhost:32000/indico:latest --indico-nginx-image=localhost:32000/indico-nginx:latest
```
Comment (arturo-seijas) [inline docs/how-to-guides/contributing.md:42]: It might be worth noting that building the images and pushing them to microk8s is needed for executing the integration tests

File: docs/how-to-guides/contributing.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+```shell
+tox --notest -e unit
+source .tox/unit/bin/activate
+```
+
+### Testing
+
+```shell
+tox -e fmt           # update your code according to linting rules
```
Comment (arturo-seijas) [inline docs/how-to-guides/contributing.md:38]: Running just `tox`and mentioning what it does under the hood might be easier to follow

File: docs/explanation/charm-architecture.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+The workload that this container is running is defined in the [indico dockerfile in the charm repository](https://github.com/canonical/indico-operator/blob/main/indico.Dockerfile).
+
+On startup, pebble is used to start a uwsgi server in http mode, on port `8081`. Non-static traffic is forwarded here from the nginx container.
+
+## Celery container
+
+The celery container runs the same workload as the indico container, as defined in the [indico dockerfile in the charm repository](https://github.com/canonical/indico-operator/blob/main/indico.Dockerfile).
+
+On startup, pebble is used to start the Celery worker. This container will be used to process tasks created by the Indico application.
+
```
Comment (arturo-seijas) [inline docs/explanation/charm-architecture.md:39]: The `prom/statsd-exporter:v0.22.8` exporter image is missing

File: docs/explanation/charm-architecture.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# Charm Architecture
+
+At it's core, Indico is a flask application that requires connections to PostgreSQL, Redis, and Celery.
+
+In designing the charm, we've leveraged the sidecar pattern to allow us to run multiple containers in each pod as follows:
+
+1. An nginx container, which can be used to efficiently serve static resources, as well as be the incoming point for all web traffic to the pod.
+2. The Indico container itself, which has a uwsgi server configured in http mode.
+3. A celery container, running tasks generated by the web application.
+4. A prometheus-nginx-exporter container which can be used to provide statistics on web traffic.
```
Comment (arturo-seijas) [inline docs/explanation/charm-architecture.md:10]: The prom/statsd-exporter:v0.22.8 exporter image is missing

File: .github/workflows/promote_charm.yaml [hunk 7]
```diff
@@ -24,3 +24,31 @@ jobs:
...
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v3
+      - name: Publish documentation
+        uses: canonical/upload-charm-docs@alpha-5
+        with:
+          discourse_host: discourse.charmhub.io
+          discourse_api_username: ${{ secrets.DISCOURSE_API_USERNAME }}
+          discourse_api_key: ${{ secrets.DISCOURSE_API_KEY }}
+          draft_mode: true
```
Comment (arturo-seijas) [inline .github/workflows/promote_charm.yaml:39]: What are the implications of `draft_mode`?

File: .github/workflows/test.yaml [hunk 8]
```diff
@@ -7,3 +7,15 @@ jobs:
   unit-tests:
     uses: canonical/operator-workflows/.github/workflows/test.yaml@main
     secrets: inherit
+  draft-publish-docs:
```
Comment (arturo-seijas) [inline .github/workflows/test.yaml:10]: I think you might want to run this as part of the `test_and_publish_charm` workflow, the one that actually publishes the charm to edge.

File: docs/how-to-guides/contributing.md (hunk 3 above)
Comment (amandahla) [reply docs/how-to-guides/contributing.md:42]: Typo:
"--indico-image="

File: docs/how-to-guides/refreshing-external-resources.md [hunk 9]
```diff
@@ -0,0 +1,10 @@
+# Refreshing external resources
+
+To refresh external resources, which includes themes and customization sources, use the `refresh-external-resources` action provided by this charm. You'll need to run this action against each unit in your application.
```
Comment (amandahla) [inline docs/how-to-guides/refreshing-external-resources.md:3]: Do you think would be helpful to add an example of the config command?
```
juju config indico customization_sources_url=https://github.com/myindico/myindico-example-files.git
```
And that the GIT repository should be reachable from the unit.

Comment (arturo-seijas) [reply docs/how-to-guides/refreshing-external-resources.md:3]: Btw, this is what we currently have:

To refresh external resources, which includes themes and customization sources, use the refresh-external-resources action provided by this charm. You’ll need to run this action against each unit in your application.

Note that if those resources include python packages from a git repository, the version number has to be incremented for the change to take effect.

For example, if you have two units (indico/0 and indico/1) you would run:
```
juju run-action --wait indico/0 indico/1 refresh-external-resources
 ```

File: docs/how-to-guides/contributing.md (hunk 2 above)
Comment (jdkandersson) [reply docs/how-to-guides/contributing.md:17]: I'll update it, for reference, this was a straight copy from the current documentation

File: docs/how-to-guides/contributing.md (hunk 3 above)
Comment (jdkandersson) [reply docs/how-to-guides/contributing.md:42]: I have added that just above the code block

File: docs/how-to-guides/contributing.md (hunk 4 above)
Comment (jdkandersson) [reply docs/how-to-guides/contributing.md:38]: I have changed the code block to a list of commands with an explanation what they do

File: docs/explanation/charm-architecture.md (hunk 6 above)
Comment (jdkandersson) [reply docs/explanation/charm-architecture.md:10]: Added

File: .github/workflows/promote_charm.yaml (hunk 7 above)
Comment (jdkandersson) [reply .github/workflows/promote_charm.yaml:39]: It would be good, the main issue is that charmhub uses the `docs` key in both cases so it would require us manually changing the `docs` entry depending on which channel is being targeted.

The action could actually help with that by keeping track of different `docs` link for a range of channels and then manipulating the `docs` in `metadata.yaml` before publishing. Alternatively, we could file an issue on an appropriate repo asking for separate `docs` keys for different channels, although this would also require charmhub to make changes and might take some time

File: docs/how-to-guides/refreshing-external-resources.md (hunk 9 above)
Comment (jdkandersson) [reply docs/how-to-guides/refreshing-external-resources.md:3]: Added the information

File: .github/workflows/promote_charm.yaml (hunk 7 above)
Comment (jdkandersson) [reply .github/workflows/promote_charm.yaml:39]: I have also added it in draft mode on the `test_and_publish_charm` workflow. Will update the `draft_mode` to `dry_run` as suggested in the other PR

Comment (jdkandersson) [reply .github/workflows/promote_charm.yaml:39]: I added an issue to `charmcraft`: https://github.com/canonical/charmcraft/issues/935

**Comment by jdkandersson** [comment]
I made a mistake switching the parameter which resulted in the documentation getting created, the upside is that it is now possible to compare the currently published documentation with what will be published by the action:

* current: https://discourse.charmhub.io/t/indico-docs-index/6552
* new: https://discourse.charmhub.io/t/indico-documentation-overview/7571

**Review by arturo-seijas — DISMISSED** [review]
LGTM

**Review by merkata — DISMISSED** [review]
Looks good, I don't have any other feedback.

**Review by weiiwang01 — APPROVED** [review]

**Review by merkata — APPROVED** [review]
Looks great!

**Review by arturo-seijas — APPROVED** [review]
LGTM
